src/lib/ledger_builder/ledger_builder.ml

```diff
@@ -1188,478 +1188,410 @@ end = struct
+          (*There's enough work to fill up first partition. Add txns to the first partition without the coinbase because we know there's atleast one bundle of work for a slot in the second parition which can be used for the coinbase if all the slots in the first partition are filled*)
+          let res =
+            one_prediff cw_seq ts_seq self x ~add_coinbase:false max_job_count
+          in
+          match Resources.available_space res with
+          | 0 ->
+              (*generate the next prediff with a coinbase at least*)
+              let max_jobs =
+                max_job_count - Sequence.length res.completed_work_rev
+              in
+              let res2 =
+                one_prediff res.discarded.completed_work
+                  res.discarded.user_commands_rev self y ~add_coinbase:true
+                  max_jobs
+              in
+              make_diff res (Some res2)
+          | x -> (
+              if Sequence.is_empty res.discarded.user_commands_rev then
+                (*There are no more user_commands to be added in the second partition and so just add one coinbase to fill an empty slot in the first partition and be done*)
+                let new_res = Resources.incr_coinbase_part_by res `One in
+                make_diff new_res None
+              else
+                match x with
+                | 1 ->
+                    (*There's a slot available in the first partition, fill it with coinbase and create another pre_diff for the slots in the second partiton with the remaining user commands and work *)
+                    let new_res = Resources.incr_coinbase_part_by res `One in
+                    let res2 =
+                      one_prediff new_res.discarded.completed_work
+                        new_res.discarded.user_commands_rev self y
+                        ~add_coinbase:false
+                        ( max_job_count
+                        - Sequence.length new_res.completed_work_rev )
+                    in
+                    make_diff new_res (Some res2)
+                | 2 ->
+                    (*There are two slots which cannot be filled using user commands, so we split the coinbase into two parts and fill those two spots*)
+                    let new_res = Resources.incr_coinbase_part_by res `Two in
+                    let cw_seq = new_res.discarded.completed_work in
+                    (*Create a second pre_diff for the remaing user_commands*)
+                    let res2 =
+                      one_prediff cw_seq new_res.discarded.user_commands_rev
+                        self y ~add_coinbase:false
+                        ( max_job_count
+                        - Sequence.length new_res.completed_work_rev )
+                    in
+                    make_diff new_res (Some res2)
+                | _ ->
+                    Logger.error logger
+                      "Tried to split the coinbase more than twice" ;
```

Comment:
should this be a failwith? or return an error result? What does this mean when this case gets hit. Under what circumstances will this happen?

Reply:
This case should not occur given that we restrict max number of provers we can buy proofs from by bundling the proofs. If it did, then it logs the message and simply create an empty diff. I don't think it should be a failwith because this case would hit if number of transaction and provers were such that, it created more than two empty slots before filling the next partition. Maybe it should simply add one coinbase (if it is possible) and not fill in the second partition instead of creating an empty diff.

Reply:
Yes can we change to add one coinbase?

Reply:
and if it's not possible we should do a `Logger.fatal` instead of a `Logger.error`, with a message explaining that our assumptions about snark work being completed is violated or something

Reply:
What I meant by "if it is possible" is if there is enough work to add a coinbase. But yeah, I can change it to `Logger.fatal`